src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -174,5 +192,39 @@ private Type GetGenericTypeArgument(IEnumerable actual)
 
             return null;
         }
+
+        private sealed class CaseInsensitiveCharComparer : IEqualityComparer<char>
+        {
+            public bool Equals(char x, char y)
+            {
+                return char.ToLower(x) == char.ToLower(y);
+            }
+
+            public int GetHashCode(char obj)
+            {
+                return char.ToLower(obj).GetHashCode();
+            }
+        }
+
+        internal sealed class UniqueItemsContstraintResult : ConstraintResult
+        {
+            internal ICollection NonUniqueItems { get; }
+
+            public UniqueItemsContstraintResult(IConstraint constraint, object actualValue, ICollection nonUniqueItems)
+                : base(constraint, actualValue, nonUniqueItems.Count == 0)
+            {
+                NonUniqueItems = nonUniqueItems;
+            }
+
+            public override void WriteAdditionalLinesTo(MessageWriter writer)
+            {
+                if (this.Status == ConstraintStatus.Failure)
+                {
+                    writer.Write("  Not unique items: ");
+                    var output = MsgUtils.FormatCollection(NonUniqueItems);
```

Comment:
```suggestion
                    var output = MsgUtils.FormatCollection(NonUniqueItems, 10);
```

Another performance suggestion - at the moment, you're building a list of all the non-unique items in a collection, then only displaying the first 10. (That's the default maximum amount shown by `FormatCollection`, from memory.) How about setting that value from the constraint - which them means in `NonUniqueItemsInternal` you can just find the first 10 non-unique items, and then stop looking. That will help in cases where you have lots of non-unique items, like the performance test in question!

Reply:
Thanks! I hadn't considered failing fast after 10 non-uniques. I've added this in to all the places in the class :)